dali/test/python/test_operator_audio_decoder.py

```diff
@@ -0,0 +1,117 @@
+from __future__ import print_function
+from __future__ import division
+from nvidia.dali.pipeline import Pipeline
+import nvidia.dali.ops as ops
+import nvidia.dali.types as types
+import scipy.io.wavfile
+import numpy as np
+import math
+
+# generate sinewaves with given frequencies,
+# add Hann envelope and store in channel-last layout
+def generate_waveforms(length, frequencies):
+  n = int(math.ceil(length))
+  X = np.arange(n, dtype=np.float32) + 0.5
+  def window(x):
+    x = 2 * x / length - 1
+    np.clip(x, -1, 1, out=x)
+    return 0.5 * (1 + np.cos(x * math.pi))
+
+  return np.sin(X[:,np.newaxis] * (np.array(frequencies) * (2 * math.pi))) * window(X)[:,np.newaxis]
+
+names = [
+  "/tmp/dali_test_1C.wav",
+  "/tmp/dali_test_2C.wav",
+  "/tmp/dali_test_4C.wav"
+]
+
+freqs = [
+  np.array([0.02]),
+  np.array([0.01, 0.012]),
+  np.array([0.01, 0.012, 0.013, 0.014])
+]
+rates = [ 16000, 22050, 12347 ]
+lengths = [ 10000, 54321, 12345 ]
+
+def create_test_files():
+  for i in range(len(names)):
+    wave = generate_waveforms(lengths[i], freqs[i])
+    wave = (wave * 32767).round().astype(np.int16)
+    scipy.io.wavfile.write(names[i], rates[i], wave)
+
+
+create_test_files()
+
+rate1 = 16000
+rate2 = 12999
+
+class DecoderPipeline(Pipeline):
+  def __init__(self):
+    super(DecoderPipeline, self).__init__(batch_size=8, num_threads=3, device_id=0,
+                                          exec_async=True, exec_pipelined=True)
+    self.file_source = ops.ExternalSource()
+    self.plain_decoder = ops.AudioDecoder(dtype = types.INT16)
+    self.resampling_decoder = ops.AudioDecoder(sample_rate=rate1, dtype = types.INT16)
+    self.downmixing_decoder = ops.AudioDecoder(downmix=True, dtype = types.INT16)
+    self.resampling_downmixing_decoder = ops.AudioDecoder(sample_rate=rate2, downmix=True, dtype = types.FLOAT)
+
+  def define_graph(self):
+    self.raw_file = self.file_source()
+    dec_plain,_ = self.plain_decoder(self.raw_file)
+    dec_res,_ = self.resampling_decoder(self.raw_file)
+    dec_mix,_ = self.downmixing_decoder(self.raw_file)
+    dec_res_mix,_ = self.resampling_downmixing_decoder(self.raw_file)
+    out = [dec_plain, dec_res, dec_mix, dec_res_mix]
+    return out
+
+  def iter_setup(self):
+    list = []
+    for i in range(self.batch_size):
+      idx = i % len(names)
+      with open(names[idx], mode = "rb") as f:
+        list.append(np.array(bytearray(f.read()), np.uint8))
+    self.feed_input(self.raw_file, list)
+
+def test_decoded_vs_generated():
```

Comment:
I would still compare with librosa to be aware how close we are just in case we need to debug some discrepancies.

Reply:
OK, I've checked. Librosa does a bad thing here - namely, they do this:
1. calculate output length as `ceil(in_length * out_rate / in_rate)`
2. resample at a slightly different effective rate: in_rate * out_length / in_length
So, they're changing the sampling rate instead of doing what is asked of them. At least it looks like it - I calculated the "corrected" (or rather librosa-like-distorted) rate and generated the new reference signal. Librosa matches this one much better than the original "ground truth".
I don't think it's our goal to reproduce this particular bug - because it is a bug; among other things, it precludes resampling in variable length chunks, because each output chunk will be resampled at a potentially different rate.

Reply:
I can add some tests with generous epsilon - or abandon the weird rate of 12999 Hz, but I don't think it does much good.
Anyway, since there's a random playback speed distortion as an augmentation (and a ton of others), I don't think anyone will look into 0.01% change in sampling rate vs some other library.

Reply:
> OK, I've checked. Librosa does a bad thing here - namely, they do this:
    calculate output length as ceil(in_length * out_rate / in_rate)

Doesn't they have an option to turn off this called `fix` ?

Reply:
And maybe we should have a switch to support this. I believe they just wanted to make sure that the calculated number of output samples is an integer number.

Reply:
we might want to leave the librosa comparing for just the typical conversions (22050 to 16000 and such)

Reply:
I don't know what happens to the rate then, but the output length was different, so I couldn't compare the arrays directly.